### ISSUE TITLE
CTC implementation, requesting help

### ISSUE THREAD

**Issue by DabiaoMa** [Microsoft/CNTK#1980]
I am not clear about how the methods ' labels_to_graph, forward_backward' work and or how they are related to CTC. The doc is too short for a beginner.

I guess the input to 'labels_to_graph' is an array of dimension (T_label, dim_label), where T_label is the length of the label, dim_label is  the possible characters plus blank; and the method  ' forward_backward' calculates the possibility(or possible paths) of the label given an output with dimension (T_feature, dim_label).

Am I right?

**Comment by hajix** [comment]
```
    graph = labels_to_graph(labels)
    ce = forward_backward(graph, network_output, 132)
    err = edit_distance_error(graph, network_output, 0, 1, 1, True, [132])
    trainer = Trainer(network_output, (ce, err), adadelta(network_output.parameters, lr), progress_writers)
```
By `forward_backward`, you'll use CTC as loss function or criterion. `edit_distance_error` is as evaluation function. Both loss and evaluation functions need some graph object as the input which made by `labels_to_graph`.

trainer object need network_output, a tuple contains criterion and evaluation functions, and a learner (look [here](https://cntk.ai/pythondocs/cntk.learners.html) for more info about cntk.lerners).

hope it help

**Comment by DabiaoMa** [comment]
@hajix 
Thank you and it really helps! While I have some more questions.   Is the 'network_output' the possibility of characters before or after softmax function? 
 Is the 'labels' a time series of vectors with entries 0 or 1?  
The first parameter of 'edit_distance_error' should be 'labels' instead of 'graph' according to [https://github.com/Microsoft/CNTK/blob/master/Tests/EndToEndTests/Speech/LSTM_CTC_MLF/lstm.bs](url)??

**Comment by hajix** [comment]
I don't use softmax, so it is output of thr network without any modification.
The labels are are generated from cntk deserializers, and are one-hot time series vectors.
And you are right about `edit_distance_error` input. Its first argument must be labels.

**Comment by DabiaoMa** [comment]
@hajix 
I did some experiments about methods ' edit_distance_error' and 'forward_backward'. It turns out that the arguments 'labels' and 'network_output' must have the same number of time series sequences.
But CTC is used to solve the problem that the #input _sequences and #out _sequences are possibly different. So what is the point of the two methods???

I really think a demonstration of how to use CTC with python needs to be provided.

**Comment by vmazalov** [comment]
@DabiaoMa, the number of sequences should be the same, but their lengths can be different. 

CTC python tutorial and implementation is already in the master branch. You can try it now if pull and build from master or wait a little for it to be in the official release.

**Comment by DabiaoMa** [comment]
@vmazalov 

Glad to hear that, but would you please tell me the location of the CTC python tutorial.... I did not find it.

**Comment by vmazalov** [comment]
https://github.com/Microsoft/CNTK/blob/master/Tutorials/CNTK_208_Speech_Connectionist_Temporal_Classification.ipynb

Keep in mind that the tutorial relies on some changes that are not in the official release yet. You can run it, if you build the latest master.

**Comment by DabiaoMa** [comment]
@vmazalov Thanks a lot

**Comment by kvcuong** [comment]
Hi,
I would to build a CRNN (CNN-LSTM-CTC) for handwritten recognition, the output of LSTM is sequences of size 512 x 32 (with 512 time-steps) and the CTC label sequence has the size of 80 (characters).

My question is it is possible to connect LSTM with CTC using CNTK ?
Thanks a lot

**Comment by delzac** [comment]
For anyone looking for a CTC Encoder to convert your labels into the ctc data format that cntk will accept you can check out my extension library [cntkx](https://github.com/delzac/cntkx), it has a `CTCEncoder` to help do just that.
